### PR TITLE
Material datum implementation

### PR DESCRIPTION
**Pull request by ComicIronic**
m_amt and g_amt (and in cases, gold_amt, diamond_amt, etc) have been stripped out, to make way for....
## MATERIAL DATUMS

Like reagents, but for materials!
Not generated on every object, but atoms can be set to start with an empty datum or with a datum with preset values via the starting_materials var
Think of all the possibilities from materials, but datums. Incredible.
Datums have been pooled where possible, to help save on memory load with fabricators making and eating items.
### Fabricators

Fixed a bug with fabricators not taking their shown cost, but just the design's plain cost.
Fixed a bug where fabricators would begin taking material costs even if an item wasn't completable.
Fabricators now transfer the materials used to their products: a more efficient fabricator will therefore use less metal to make a mech, but also make a mech that contains less metal.
All designs are now world-consistent - scanning the same object will ref the same design, all fabricators build off the same designs, etc.
Mechanic fabs now add the plastic cost to designs themselves, instead of it being a part of the design. This is toggled by var, and autolathes do not do this.
Mechanic design making changed to not randomly generate design costs based on tech level now. It may be resurrected later for the fabs themselves.
Mechanic blueprints no longer come with uses. Instead, the blueprint will be consumed if paper, or not consumed if nano, when put into a machine.
Mechanic designs are now material-specific: it is possible to now scan a specific copy of an item if its materials are unique, and then recreate the item with those materials by using the general fabricator.
### Autolathe

Moved to be a research type fabricator under mechanic_fabs.
Autolathe building has a maximum time to bring it in line with older behaviour, especially since the lathe's design list never changes.
Autolathe does not require plastic.
### Wiring

Moved RnD wiring under a rnd wiring datum, which is also used for the Autolathe hacking.
# UP FOR CODE REVIEW
### TODO
- [x] Bring back autolathe recycling
- [x] Bring back autolathe hacking
- [x] Finalise autholathe code
- [x] SQUASH
### TO DISCUSS
- Should you be able to add designs to autolathes? These designs would be the same as those in the general fabricator - currently not possible.

### CODE REVIEW

**Comment by Iamgoofball** [comment]
>129 files changed

**Comment by ComicIronic** [comment]
m_amt and g_amt dig deep.
Whichever oldcoder made those vars was clearly a terrible human being, as is everyone who supported them.

**Comment by 9600bauds** [comment]
You absolute madman
In the future, will fabricator designs have their cost depend on the material datums of each item?

**Comment by ririchiyo** [comment]
this is retarded stop adding shitty lists to things that don't need it.

**Comment by 9600bauds** [comment]
>this is retarded

that's not a very good review

**Comment by clusterfack** [comment]
BUT DOES IT LAG

**Comment by ComicIronic** [comment]
Oh you better believe it lags. The server doesn't even run anymore, you just stare at the splash screen and imagine the round.
<s>It hasn't had any noticable impact because the hooks are very spread out</s>

**Comment by PJB3005** [comment]
Does this mean goof ores can be added back in?

hype?

**Comment by ComicIronic** [comment]
Goof ores will never be added back in.
Someone post that image of the Goof clown asteroid.

**Comment by Iamgoofball** [comment]
@ComicIronic i had it saved somewhere but I lost it, RIP

if anything don't add them back in as ores holy shit

make them cargo orderables or some shit idfk

**Comment by 9600bauds** [comment]
goofore
never forgetti

**Comment by Iamgoofball** [comment]
neverforgetti should be a dish

**Comment by Probe1** [comment]
>Should mechanic design uses be removed? They are kinda pointless.

Might as well since no one uses them when nanopaper is plentifully available.  I feel as though mechanics are in a really good place right now with the current bugs and wonkyness.  People don't scan and load every single thing under the sun.  They just scan what they predict they will print multiple items of and let it rip.  Having limited uses doesn't play into that.  

As long as there is a choice between unlimited, and limited, and the choice costs the same then people will choose unlimited.

**Comment by ComicIronic** [comment]
Alright.

Instead, nanoprints will not delete on use, but paperprints will.

**Comment by Probe1** [comment]
Hold on I thought you were talking about paper prints in what I quoted.  I'm confused.

**Comment by ComicIronic** [comment]
I was.
I'm removing the use number system, because it's pointless and redundant.
To keep the difference between paperprints and nanoprints, paperprints are now consumed and nanoprints are not.

**Comment by Probe1** [comment]
Okay that's fine.  I think the core issue is the only people that use paper prints are ones that have not yet realized nanoprints exist yet.  You could rename them Napoleonic Sheets and make them play oggs when used and they still wouldn't have much of a point.

**Comment by 9600bauds** [comment]
One thing I've been meaning to look into: Maybe after every item is made to have a list of the materials that it's made of, you could have the reverse engine/device analyzer be able to scan items without tech levels?
So you could, for example, mass produce desk lamps or clown masks.

**Comment by Probe1** [comment]
how does he do it?  what a wunderkind.  he made 9 commits disappear into 1.

**Comment by Probe1** [comment]
After reviewing, I oppose the changes to the general fabricator.  Items now cost as much as the autolathe IN ADDITION to costing plastic.  I had a lot more uninteresting arguments to make but I found a bug.  When you try to deconstruct an autolathe it does not deconstruct it but it does make a machine frame appear and what's stranger, it's the machine frame for a general fabricator complete with its boards and parts.

ss 2015-05-24 at 10 59 43

**Comment by Probe1** [comment]
Crazy bug aside, it now costs more even with T3 upgrades in a genfab to print something than it does to just use the autolathe.  Why would anyone want to track an item down, scan it, reverse engine it, print it, add it, add the materials.. just to pay more?

:-1:

**Comment by ComicIronic** [comment]
It can still build items that neither the autolathe or protolathe can. Replacing those machines was never its intended purpose.

The big change the gen fab gets is the ability to recreate items with their materials. Since I'm planning to expand the materials system, the gen fab will be the only way to mass produce custom items, as well as the only way to produce items with no design.

**Comment by Probe1** [comment]
I still strongly believe that the current place of the genfab is to be able to mass produce items with existing designs cheaper and more efficiently.  Right now without autolathe queuing there is no good way to say, make 500 wrenches.  But with a general fabricator I can.  Sure I have to print a wrench, scan a wrench, reverse engine a wrench, print a wrench, use plastic and other materials to make a wrench but *once I meet all the requirements* I can cheaply make tons of wrenches.

Equalizing the genfab design costs to the autolathe and giving the autolathe queuing negates both aspects of the general fabricators current usage.  I'm a-ok with making the autolathe less clunky and giving it a giant QoL improvement by adding queueing.  However that does negate half the argument for the general fab.

When you remove it producing things more cheaply but at a cost of plastic and replace it with costing the same price as an autolathe **PLUS an additional plastic cost** you haven't just removed it from the niche it filled, you've made it even worse than an autolathe or protolathe.  

My solitary issue with it was the exploitability of rcd_ammo when used in conjunction with a material synth and I feel rcd_ammo specifically should be increased from an average of 1500cm3 to 8000cm3 materials (total).  Here, we move from 1500 to 40,000.  It nerfs it out of viability.  Even with T3 parts, the additional cost of **plastic** makes it as expensive as an autolathe.

The unconstructable items are limited, especially after the majority of them were removed from the list after the initial implementation of Mechanics led to it becoming the Gun and Baton Shop.  

I believe that the general fabricator currently fills a role and does it very well.  You have to put in an upfront investment of time and use a rare resource but you get back items easier and cheaper than you would otherwise.  Removing both the easier aspect by giving autolathes queuing as well as increasing the costs of the items produced removes both aspects of the general fab that made it a welcomed addition.

**Comment by 9600bauds** [comment]
I honestly have to step in right here and agree with Probe1. The idea of Mechanics was once to be that of "Science does the research, Mechanics do the mass-producing and distribution".
This right here is kind of completely killing the gen fab.

**Comment by Probe1** [comment]
I believe when mechanics were introduced the idea was someone fixes and upgrades the machines which scientists never did.  

What it turned into is someone fixes and upgrades the machines and also produces all the RnD items and distributes them to the crew.  

Scientists have proven over the entire 11 months I've been here that they are not generally interested in standing in RnD and making people stuff.  There are of course exceptions.  Mechanics on the other hand exist to build people stuff.  The exception to mechanic is that rare fig that puts up rwalls in maint and takes over aux surgery so he can build an autism cave with every vending machine and never leave.

**Comment by ComicIronic** [comment]
>"Science does the research, Mechanics do the mass-producing and distribution".

Which is still true, and not something you have to tell me because I coded the job from the ground up.

>This right here is kind of completely killing the gen fab.

No, it's killing the gen fab's current function, which is entirely intended because the gen fab was never intended to do anything other than build items that are otherwise unbuildable, and it's going to do that very well.

What the gen fab became is all due to the inadequacies of the old autolathe code.

**Comment by 9600bauds** [comment]
If it's still true, it's a playerbase problem and not job design, since Mechanics lose anything that sets them apart from Science for mass-producing or otherwise.
Hopefully making more common items build-able will make the Gen Fab do a comeback, but... really, why does it have a plastic cost on top of the FULL material cost?

**Comment by Probe1** [comment]
I have no idea what items you're envisioning.  The only thing that comes to mind are flashes.  The majority of things can either be printed from the protolathe, autolathe, or circuit imprinter, or are unscannable, or are undesirable and plentiful enough already.

The general fabricator is a wonderful second stage production machine.  That means that it isn't where you produce the item for the first time, it's where you mass produce it.  It uniquely fits this role currently.

Let's say, let's say we're mechanics.  And medbay wants bluespace beakers.  They're about 5000 metal a pop.  No big deal that's like 2 sheets each.  So first we research bluespace beakers.  Then we print a single one.  And scan it.  And reverse engine it.  And load the printer and print it. And load it.  Then we can get a good discount on producing them in bulk with the genfab.  

It's extremely rare that someone uses the general fabricator as a first step production machine by scanning and loading normal items on the station.  There are examples like airlock controller boards and handheld teleporters.  But they're extremely limited.  And other coders (like Kurfursts excellent soldering iron) work towards reducing that need of making the items anyways.

I think how it works right now is fantastic and it shouldn't be nerfed to only viably producing a handful of items because they could not be made any other way.  Right now it's a second stage production machine that lets you produce something more expensively but later on cheaper as long as you have a supply of plastic.  That's an awesome machine.  The flatpack fabricator is a good example of where you want to go with the genfab.  It's total shit.  It's more expensive and takes longer to build machines that it would by using the protolathe and circuit imprinter.  It's only areas of usefulness are when you have an unrealistic amount of resources or it's at a specific time in the round where it is faster to scan an existing machine and print it than it is to wait on research to finish RnD so you can print it normally.

The flatpack fabricator is a "pay more for less" machine and with this update, so will be the general fabricator.

**Comment by Probe1** [comment]
Anyway I think the general fab is great how it is.  It isn't what you envisioned but it's something that we can use and fulfills a role we didn't have.  It's multidimensional.  It can produce already made things cheaper and it can produce some things you can't get elsewhere.  The protolathe has it right there in the name - proto.  First.  General fab is for once you already have stuff.  

It'd be a disservice to how well it turned out to cut out half of what it's used for.

**Comment by ComicIronic** [comment]
> The protolathe has it right there in the name - proto. First. General fab is for once you already have stuff. 

And that will be its eventual purpose again, because it will become the only way to mass-produce items of any specific material makeup, but it's current function is just a result of mistakes made in its design and implementation.

The general fab will have a niche, but it won't be its current one because that one was a bad idea and lacked good consistency with the rest of the game.

When my planned changes to research come in, you'll see the genfab get a lot more use, hopefully.

**Comment by Probe1** [comment]
Where you see inconsistency, I see novelty.  Most of the other mechanics I encounter see it as a unique opportunity as well.

The flatpack fabricator is a bad design.  It costs more than assembling by hand and takes longer.   This will make the general fabricator the same story.  More work and more expensive to boot.

  I'll still play mechanic but it will remove most of the interaction I have with Mechanic machines.  I don't know what you're envisioning for it but we already have one.  It's a good one.  I wish you wouldn't nerf it because it's not what you wanted.  The playerbase is happy with it and this will change it from a unique machine to a snowflake machine.  It needs a few specific items rebalanced, not nerfed to oblivion.

That's my final opinion on it.

**Comment by 9600bauds** [comment]
Good autolathe when

**Comment by ComicIronic** [comment]
A henk henk henk henk henk henk never

**Comment by 9600bauds** [comment]
But it's so good

**Comment by 9600bauds** [comment]
This when

**Comment by Probe1** [comment]
>Fixed a bug with fabricators taking design cost, not shown cost.

not a bug and introduces design oversights.

**Comment by ComicIronic** [comment]
Yes a bug.

>fabricator says design is 30000 metal
>takes 20000 metal
>upgrade fabricator
>fabricator says designs is 10000 metal
>takes 20000 metal

**Comment by Probe1** [comment]
plus plastic.  everything will cost the same as an autolathe making scanning it pointless in the first place AND cost plastic on top of it.  I still don't understand what your aim is other than removing the general fab from use.

**Comment by 9600bauds** [comment]
that's another deal though

**Comment by 9600bauds** [comment]
NOW

**Comment by Probe1** [comment]
rip genfab

**Comment by 9600bauds** [comment]
fabolathe rise

File: code/game/machinery/autolathe.dm [hunk 1]
```diff
@@ -1,10 +1,40 @@
...
+
+	build_time = AUTOLATHE_BUILD_TIME
+
+	removable_designs = 0
+	plastic_added = 0
+
+	allowed_materials = list(
+						MAT_IRON,
+						MAT_GLASS
+	)
```
Comment (Probe1) [inline code/game/machinery/autolathe.dm]: why the 6 tabs?

**Pull request merged**